examples/fabric-admin/commands/common/StayActiveSender.cpp

```diff
@@ -0,0 +1,83 @@
+/*
+ *    Copyright (c) 2024 Project CHIP Authors
+ *    All rights reserved.
+ *
+ *    Licensed under the Apache License, Version 2.0 (the "License");
+ *    you may not use this file except in compliance with the License.
+ *    You may obtain a copy of the License at
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *    Unless required by applicable law or agreed to in writing, software
+ *    distributed under the License is distributed on an "AS IS" BASIS,
+ *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *    See the License for the specific language governing permissions and
+ *    limitations under the License.
+ */
+
+#include "StayActiveSender.h"
+
+#include <app-common/zap-generated/cluster-objects.h>
+#include <app/ConcreteCommandPath.h>
+#include <controller/InvokeInteraction.h>
+#include <support/CHIPMem.h>
+
+StayActiveSender::StayActiveSender(uint32_t stayActiveDuration, const chip::ScopedNodeId & peerNode,
+                                   chip::app::InteractionModelEngine * engine, OnDoneCallbackType onDone) :
+    mStayActiveDuration(stayActiveDuration),
+    mPeerNode(peerNode), mpImEngine(engine), mOnDone(onDone), mOnConnectedCallback(HandleDeviceConnected, this),
+    mOnConnectionFailureCallback(HandleDeviceConnectionFailure, this)
+{}
+
+CHIP_ERROR StayActiveSender::SendStayActiveCommand(chip::Messaging::ExchangeManager & exchangeMgr,
+                                                   const chip::SessionHandle & sessionHandle)
+{
+    auto onSuccess = [&](const chip::app::ConcreteCommandPath & commandPath, const chip::app::StatusIB & status,
+                         const auto & dataResponse) {
+        uint32_t promisedActiveDuration = dataResponse.promisedActiveDuration;
+        ChipLogProgress(ICD, "StayActive command succeeded with promised duration %u", promisedActiveDuration);
+        mOnDone(promisedActiveDuration);
+        chip::Platform::Delete(this);
```

Comment:
Is it guaranteed that this class is created with Platform::New?
Could we ensure this by making the constructor private and have a `Create` method instead, so that both New and Delete are always together in the same file?

Reply:
Almost. Unfortunately `Platform::New` is not able to call private function :(

```
2024-08-13 20:08:03 INFO    ../../examples/fabric-admin/third_party/connectedhomeip/src/lib/support/CHIPMem.h: In instantiation of ‘T* chip::Platform::New(Args&& ...) [with T = StayActiveSender; Args = {unsigned int&, const chip::ScopedNodeId&, chip::app::InteractionModelEngine*, std::function<void(unsigned int)>&}]’:
2024-08-13 20:08:03 INFO    ../../examples/fabric-admin/commands/common/StayActiveSender.cpp:29:66:   required from here
2024-08-13 20:08:03 INFO    ../../examples/fabric-admin/third_party/connectedhomeip/src/lib/support/CHIPMem.h:148:16: error: ‘StayActiveSender::StayActiveSender(uint32_t, const chip::ScopedNodeId&, chip::app::InteractionModelEngine*, OnDoneCallbackType)’ is private within this context
2024-08-13 20:08:03 INFO      148 |         return new (p) T(std::forward<Args>(args)...);
2024-08-13 20:08:03 INFO          |                ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
2024-08-13 20:08:03 INFO    ../../examples/fabric-admin/commands/common/StayActiveSender.cpp:40:1: note: declared private here
2024-08-13 20:08:03 INFO       40 | StayActiveSender::StayActiveSender(uint32_t stayActiveDurationMs, const chip::ScopedNodeId & peerNode,
```

Reply:
Let me know if the workaround is acceptable to you. I don't have a good alternative